lib/auth/session_access.go

```diff
@@ -0,0 +1,293 @@
+			case "roles":
+				var roles []string
+				for _, role := range ctx.Roles {
+					roles = append(roles, role.GetName())
+				}
+
+				return roles, nil
+			}
+		}
+	}
+
+	return nil, trace.NotFound("%v is not defined", strings.Join(fields, "."))
+}
+
+func (ctx *SessionAccessContext) GetResource() (types.Resource, error) {
+	return nil, trace.BadParameter("resource unsupported")
+}
+
+func (e *SessionAccessEvaluator) matchesPredicate(ctx *SessionAccessContext, require *types.SessionRequirePolicy, allow *types.SessionJoinPolicy) (bool, error) {
+	if !e.matchesKind(require.Kinds) || !e.matchesKind(allow.Kinds) {
+		return false, nil
+	}
+
+	parser, err := services.NewWhereParser(ctx)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+
+	ifn, err := parser.Parse(require.Filter)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+
+	fn, ok := ifn.(predicate.BoolPredicate)
+	if !ok {
+		return false, trace.BadParameter("unsupported type: %T", ifn)
+	}
+
+	return fn(), nil
+}
+
+func (e *SessionAccessEvaluator) matchesJoin(allow *types.SessionJoinPolicy) bool {
+	if !e.matchesKind(allow.Kinds) {
+		return false
+	}
+
+	for _, requireRole := range e.roles {
+		for _, allowRole := range allow.Roles {
+			expr := utils.GlobToRegexp(requireRole.GetName())
+			// GlobToRegexp makes sure this is always a valid regexp.
+			matched, _ := regexp.MatchString(expr, allowRole)
+
+			if matched {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func (e *SessionAccessEvaluator) matchesKind(allow []string) bool {
+	if contains(allow, e.kind) || contains(allow, "*") {
+		return true
+	}
+
+	return false
+}
+
+// CanJoin returns the modes a user has access to join a session with.
+// If the list is empty, the user doesn't have access to join the session at all.
+func (e *SessionAccessEvaluator) CanJoin(user SessionAccessContext) ([]types.SessionParticipantMode, error) {
+	supported, err := e.supportsSessionAccessControls()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if !supported {
+		return preAccessControlsModes(e.kind), nil
+	}
+
+	var modes []types.SessionParticipantMode
+
+	for _, allowPolicy := range getAllowPolicies(user) {
+		if e.matchesJoin(allowPolicy) {
+			for _, modeString := range allowPolicy.Modes {
+				mode := types.SessionParticipantMode(modeString)
+				if !SliceContainsMode(modes, mode) {
+					modes = append(modes, mode)
+				}
+			}
+		}
+	}
+
+	return modes, nil
+}
+
+func SliceContainsMode(s []types.SessionParticipantMode, e types.SessionParticipantMode) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+	}
+	return false
+}
+
+// PolicyOptions is a set of settings for the session determined by the matched require policy.
+type PolicyOptions struct {
+	TerminateOnLeave bool
+}
+
+// FulfilledFor checks if a given session may run with a list of participants.
+func (e *SessionAccessEvaluator) FulfilledFor(participants []SessionAccessContext) (bool, PolicyOptions, error) {
```

Comment:
I think its worth adding detailed comments to the body of this and the other key access evaluator methods. Given how critical this logic is, I think its best to over-comment.

Reply:
done